Source/WebSocket/websocket.cpp

```diff
@@ -0,0 +1,245 @@
+    ) HC_NOEXCEPT
+try 
+{
+    if (websocket == nullptr || headerName == nullptr || headerValue == nullptr)
+    {
+        return HC_E_INVALIDARG;
+    }
+    RETURN_IF_WEBSOCKET_CONNECT_CALLED(websocket);
+
+    websocket->connectHeaders[headerName] = headerValue;
+
+    HC_TRACE_INFORMATION(WEBSOCKET, "HCWebSocketSetHeader [ID %llu]: %s=%s",
+        websocket->id, headerName, headerValue);
+    return HC_OK;
+}
+CATCH_RETURN()
+
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketSetFunctions(
+    _In_opt_ HC_WEBSOCKET_MESSAGE_FUNC messageFunc,
+    _In_opt_ HC_WEBSOCKET_CLOSE_EVENT_FUNC closeFunc
+    ) HC_NOEXCEPT
+try 
+{
+    auto httpSingleton = get_http_singleton();
+    httpSingleton->m_websocketMessageFunc = messageFunc;
+    httpSingleton->m_websocketCloseEventFunc = closeFunc;
+
+    return HC_OK;
+}
+CATCH_RETURN()
+
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketConnect(
+    _In_z_ PCSTR uri,
+    _In_ HC_WEBSOCKET_HANDLE websocket,
+    _In_ HC_WEBSOCKET_CONNECT_INIT_ARGS args
+    ) HC_NOEXCEPT
+try 
+{
+    auto httpSingleton = get_http_singleton();
+
+    auto connectFunc = httpSingleton->m_websocketConnectFunc;
+    if (connectFunc != nullptr)
+    {
+        try
+        {
+            connectFunc(uri, websocket, args);
+        }
+        catch (...)
+        {
+            HC_TRACE_ERROR(WEBSOCKET, "HCWebSocketConnect [ID %llu]: failed", websocket->id);
+        }
+    }
+
+    return HC_OK;
+}
+CATCH_RETURN()
+
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketSendMessage(
+    _In_ HC_WEBSOCKET_HANDLE websocket,
+    _In_z_ PCSTR message
+    ) HC_NOEXCEPT
+try
+{
+    auto httpSingleton = get_http_singleton();
+
+    auto sendFunc = httpSingleton->m_websocketSendMessageFunc;
+    if (sendFunc != nullptr)
+    {
+        try
+        {
+            sendFunc(websocket, message);
+        }
+        catch (...)
+        {
+            HC_TRACE_ERROR(WEBSOCKET, "HCWebSocketSendMessage [ID %llu]: failed", websocket->id);
+        }
+    }
+
+    return HC_OK;
+}
+CATCH_RETURN()
+
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketClose(
+    _In_ HC_WEBSOCKET_HANDLE websocket
+    ) HC_NOEXCEPT
+try 
+{
+    auto httpSingleton = get_http_singleton();
+
+    auto closeFunc = httpSingleton->m_websocketCloseFunc;
+    if (closeFunc != nullptr)
+    {
+        try
+        {
+            closeFunc(websocket);
+        }
+        catch (...)
+        {
+            HC_TRACE_ERROR(WEBSOCKET, "HCWebSocketClose [ID %llu]: failed", websocket->id);
+        }
+    }
+
+    return HC_OK;
+}
+CATCH_RETURN()
+
+HC_WEBSOCKET_HANDLE HCWebSocketDuplicateHandle(
+    _In_ HC_WEBSOCKET_HANDLE websocket
+    ) HC_NOEXCEPT
+try 
+{
+    if (websocket == nullptr)
+    {
+        return nullptr;
+    }
+
+    HC_TRACE_INFORMATION(WEBSOCKET, "HCWebSocketDuplicateHandle [ID %llu]", websocket->id);
+    ++websocket->refCount;
+
+    return websocket;
+}
+CATCH_RETURN_WITH(nullptr)
+
+HC_API HC_RESULT HC_CALLING_CONV
+HCWebSocketCloseHandle(
```

Comment:
>HCWebSocketCloseHandle [](start = 0, length = 22)

websocket close is talk to service and perform a close communication if possible. I don't think you can close multiple times.